sender/nats.go

```diff
@@ -0,0 +1,171 @@
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation; either
+ * version 2.1 of the License, or (at your option) any later version.
+ *
+ * This library is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public
+ * License along with this library; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA
+ * 02110-1301  USA
+ */
+
+ package sender
+
+ import (
+	"fmt"
+	"github.com/nats-io/nats.go"
+	"github.com/telenornms/skogul"
+	"github.com/telenornms/skogul/encoder"
+	"sync"
+	"crypto/tls"
+)
+
+var natsLog = skogul.Logger("sender", "nats")
+/*
+Nats.io sender. A small Nats, non-jetstream, publisher with:
+
+- Authentication: Username/Password, TLS
+- Authorization: Username/Password, UserCredentials/JWT
+
+*/
+
+type Nats struct {
+	Servers		string   `doc:"Comma separated list of nats URLs"`
+	Subject		string   `doc:"Subject to publish messages on"`
+	Name		string	 `doc:"Client name"`
+	Username	string   `doc:"Client username"`
+	Password	string	 `doc:"Client password"`
+	TLSClientKey    string   `doc:"TLS client key file path"`
+	TLSClientCert   string   `doc:"TLS client cert file path"`
+	TLSCACert	string   `doc:"CA cert file path"`
+	UserCreds       string   `doc:"Nats credentials file path"`
+	NKeyFile        string   `doc:"Nats nkey file path"`
+	Insecure	bool	 `doc:"TLS InsecureSkipVerify"`
+	Encoder		skogul.EncoderRef
+	o		*[]nats.Option
+	nc		*nats.Conn
+	once		sync.Once
+}
+
+// Verify configuration
+func (n *Nats) Verify() error {
+        if n.Subject == "" {
+                return skogul.MissingArgument("Subject")
+        }
+
+        return nil
+}
+
+
+func (n *Nats) init() {
+
+	if n.Encoder.Name == "" {
+		n.Encoder.E = encoder.JSON{}
+	}
+
+	if n.Name == "" {
+		n.Name = "skogul"
+	}
+	n.o = &[]nats.Option{nats.Name(n.Name)}
+
+	if n.Servers == "" {
+		n.Servers = nats.DefaultURL
+	}
+
+	//User Credentials, use either.
+	if n.UserCreds != "" && n.NKeyFile != "" {
+		natsLog.Fatal("Please configure usercreds or nkeyfile.")
+	}
+	if n.UserCreds != "" {
+		*n.o = append(*n.o, nats.UserCredentials(n.UserCreds))
+	}
+
+	//Plain text passwords
+	if n.Username != "" && n.Password != "" {
+		if n.TLSClientKey != "" {
+			natsLog.Warnf("Using plain text password over a non encrypted transport!")
+		}
+		*n.o = append(*n.o, nats.UserInfo(n.Username, n.Password))
+	}
+
+	//TLS authentication, Note: Fix selfsigned certificates.
+	if n.TLSClientKey != "" && n.TLSClientCert != "" {
+		cert, err := tls.LoadX509KeyPair(n.TLSClientCert, n.TLSClientKey)
+		if err != nil {
+			natsLog.Fatalf("error parsing X509 certificate/key pair: %v", err)
```

Comment:
Fatal/killing on failure is not ideal. For setups with multiple pipelines, it would cause serious issues. It could mean that a skogul instance with 3 unrelated receiver-sender pipelines fail to start because just one of them fail.

If possible, I suggest storing the error on n.init_error = err, then in Send() checking if n.init_error exists and return that. It will be slightly less than ideal for setups with just a single receiver and sender, but the log messages should still be obvious.